### DISCUSSION TITLE
Sort Orders

### DISCUSSION THREAD

**Discussion by daxpryce** [microsoft/graspologic#630] [General]
In #628 @vmdhhh sorted all the imports after I asked him to sort the `from typing ...` line. He wasn't sure he should have and was a bit alarmed and I realized that while we (meaning JHU and Microsoft) had talked about this in the past, we didn't really codify what we expected.

Rather than keep expecting the outside world to know our little enclave's rules of thumb, we should write them down and put them in our contribution guide.

@bdpedigo @j1c @asaadeldin11 @alyakin314 @Nyecarr, I'm kind of initially proposing (without any super deep thought, just as a starting point for discussion):

Imports should be grouped by type, and alphabetized within that group

1. Standard lib imports group
2. Library dependencies imports
3. Graspologic imports

So an example would be:

```python
import logging
import math
import time
from typing import Any, Dict, List, Optional, Tuple


import networkx as nx
import numpy as np

from sklearn.manifold import TSNE	
import umap

from ..embed import node2vec_embed
from ..partition import leiden
from ..preprocessing import cut_edges_by_weight, histogram_edge_weight
from .classes import NodePosition
from .nooverlap import remove_overlaps
```

There's a little bit in the relative imports that make this not super straight forward - namely, should we treat `..` and `.` as equivalent for the purposes of sorting, such that it becomes:

```python
from .classes import NodePosition
from ..embed import node2vec_embed
from .nooverlap import remove_overlaps
from ..partition import leiden
from ..preprocessing import cut_edges_by_weight, histogram_edge_weight
```

Or should we keep them separated?

(Aside: is it even possible to go up more than one level with the relative imports?  Like, could I do `from ...foo import bar` from something 5 packages deep?, presuming foo.py was a toplevel module?)

**Comment by nicaurvi** [comment]
I think that this is good to codify and I agree with the way it is formatted above except that there should not be whitespace between numpy and sklearn.manifold -- they are all library dependencies.

**Reply by daxpryce** [reply]
Agreed, fixing it and showing it for others:

```python
import logging
import math
import time
from typing import Any, Dict, List, Optional, Tuple

import networkx as nx
import numpy as np
from sklearn.manifold import TSNE	
import umap

from ..embed import node2vec_embed
from ..partition import leiden
from ..preprocessing import cut_edges_by_weight, histogram_edge_weight
from .classes import NodePosition
from .nooverlap import remove_overlaps
```

**Comment by bdpedigo** [comment]
hello world

**Comment by bdpedigo** [comment]
i think i agree with the above, i always just use the `sort imports` command in VS code which I believe does exactly what was proposed.

**Reply by daxpryce** [reply]
I think IntelliJ does the same - @nyecarr, you know if I'm talking crazy, or what?  I think the only thing intellij doesn't do is order the `from foo import fizz, buzz, sandwich`

**Reply by nicaurvi** [reply]
Correct -- just doesn't sort after import using the above syntax. I know there are other plugins available in the marketplace and I'd be surprised if it wasn't supported.
